src/realm/index_string.cpp

```diff
@@ -357,46 +358,44 @@ void IndexArray::index_string_all_ins(StringData value, IntegerColumn& result, C
     struct WorkItem {
         const char* header;
         size_t string_offset;
-        int permutation; // -1 == unexpanded, otherwise 0-15
+        key_type key;
     };
     using WorkList = std::vector<WorkItem>;
     WorkList work_list;
 
-    const char* top_header = get_header_from_data(m_data);
-    work_list.push_back({top_header, 0, -1});
-
     const util::Optional<std::string> upper_value = case_map(value, true);
     const util::Optional<std::string> lower_value = case_map(value, false);
 
+    auto generate_search_keys = [&work_list, &upper_value, &lower_value](const char* header,
+                                                                         size_t string_offset) -> void {
+        std::set<key_type> keys_seen;
```

Comment:
`std::set` doesn't seem like the best choice for this scenario since we don't care about the order of the elements and so there's no reason to pay for the extra memory allocations of this tree-based structure. `std::unordered_set` is a better fit. Since we're dealing with a collection of at most 16 items I suspect that `std::vector` would give it a run for its money.